src/main/java/ch/njol/skript/expressions/ExprSignText.java

```diff
@@ -32,145 +26,203 @@
-		block = (Expression<Block>) exprs[exprs.length - 1];
-		return true;
-	}
-	
+	private Expression<Block> exprBlock;
+	private boolean delayed;
+
 	@Override
-	public boolean isSingle() {
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		isLines = (matchedPattern == 0);
+		if (!isLines)
+			line = matchedPattern == 1 ? (Expression<Number>) exprs[0] : new SimpleLiteral<>(parseResult.mark, false);
+		exprBlock = (Expression<Block>) exprs[exprs.length - 1];
+		delayed = isDelayed == Kleenean.TRUE;
 		return true;
 	}
-	
-	@Override
-	public Class<? extends String> getReturnType() {
-		return String.class;
-	}
-	
+
 	@Override
 	@Nullable
-	protected String[] get(final Event e) {
-		final Number l = line.getSingle(e);
-		if (l == null)
-			return new String[0];
-		final int line = l.intValue() - 1;
-		if (line < 0 || line > 3)
-			return new String[0];
-		if (getTime() >= 0 && block.isDefault() && e instanceof SignChangeEvent && !Delay.isDelayed(e)) {
-			return new String[] {((SignChangeEvent) e).getLine(line)};
+	protected String[] get(Event event) {
+		Block block = exprBlock.getSingle(event);
+		if (isLines) {
+			if (block == null || !sign.isOfType(block))
+				return new String[0];
+			if ((event instanceof SignChangeEvent && getTime() >= 0 && block.equals(((SignChangeEvent) event).getBlock()) && !delayed))
+				return ((SignChangeEvent) event).getLines();
+			return ((Sign) (block.getState())).getLines();
+		} else {
+			Number lineNumber = line.getSingle(event);
+			if (lineNumber == null || lineNumber.intValue() < 1 || lineNumber.intValue() > 4)
+				return new String[0];
+			int index = lineNumber.intValue() - 1;
+			if (event instanceof SignChangeEvent && getTime() >= 0 && block == ((SignChangeEvent) event).getBlock() && !delayed) {
+				return CollectionUtils.array(((SignChangeEvent) event).getLine(index));
+			}
+			if (block == null || !sign.isOfType(block)) {
+				return new String[0];
+			}
+			return CollectionUtils.array(
+				((Sign) (block.getState())).getLine(index)
+			);
 		}
-		final Block b = block.getSingle(e);
-		if (b == null)
-			return new String[0];
-		if (!sign.isOfType(b))
-			return new String[0];
-		return new String[] {((Sign) b.getState()).getLine(line)};
 	}
-	
-	@Override
-	public String toString(final @Nullable Event e, final boolean debug) {
-		return "line " + line.toString(e, debug) + " of " + block.toString(e, debug);
-	}
-	
-	// TODO allow add, remove, and remove all (see ExprLore)
+
 	@Override
 	@Nullable
 	public Class<?>[] acceptChange(final ChangeMode mode) {
-		if (mode == ChangeMode.DELETE || mode == ChangeMode.SET)
-			return new Class[] {String.class};
+		if (isLines) {
+			if (mode == ChangeMode.RESET) {
+				return null;
+			}
+			return CollectionUtils.array(String[].class);
+		} else if (mode == ChangeMode.SET || mode == ChangeMode.DELETE) {
+			return CollectionUtils.array(String.class);
+		}
 		return null;
 	}
```

Comment:
Use a switch statement to properly display the change modes that are accepted for readability and for future enum additions to changemode enum